### PR TITLE
Specialize Optional[T] to T (or subtype for Tensor) or None when executing graph

### PR DESCRIPTION
**Pull request by t-vi**
This patch specializes `Optional[Tensor]` graph inputs to either a `DimensionedTensorType` (if a Tensor is passed) or `NoneType`. Other `Optional[T]` are specialized to `T` or `None`.

## Functional changes
- For unwrapping (checked and unchecked) we need to keep the output type, as IR code that follows unwrapping may not work with NoneType (just as it doesn't deal with Optional). While it would not be hit during execution, it will run against the (legitimate) assumptions of the analysis passes.
- Function lookup currently will not match NoneType when it expects optional (I'm not entirely sure why this doesn't lead to unhappyness currently, but hey), I amend this at the level of the function matching code (`operator.cpp`), but see Adam's comments. We would run into trouble if we needed to select between functions whose signature only differs in Optional types with different subtypes, but we would have the same problem when calling them directly, so I would think this is OK.

## Advantages:
- It would enable throwing away branches we can't hit. This also reduces the "blockyness" of the graph, so it may be easier to apply optimizations (e.g. fuse things in `if t is None: ...` and outside the `if`.
- Arguments passed into `Optional[Tensor]` arguments will get shape information, which is very handy.
- It get's rid of the problem that tensors passed into Optional arguments get requires_grad set erroneously #18270 (though that also affects lists, which aren't fixed here).
- `Optional[List[int]]` is needed for #18697.

## Disadvantages:
- We're changing typing in a more subtle way than the `TensorType`->`DimensionedTensorType`.
- In particular, specializing to NoneType loses the Type information captured in the `OptionalType` element type.

### CODE REVIEW

File: torch/csrc/jit/fuser/fallback.cpp
Comment (apaszke) [inline torch/csrc/jit/fuser/fallback.cpp]: Please remove!

File: torch/csrc/jit/passes/shape_analysis.cpp
Comment (apaszke) [inline torch/csrc/jit/passes/shape_analysis.cpp]: This comment is messed up now

File: torch/csrc/jit/operator.cpp
Comment (apaszke) [inline torch/csrc/jit/operator.cpp]: Uh how come do we need this special case here, and how did we survive without it before? We should be able to unify a `NoneType` with an `OptionalType`, right?

File: torch/csrc/jit/passes/shape_analysis.cpp
Comment (apaszke) [inline torch/csrc/jit/passes/shape_analysis.cpp]: Is this just formatting? Can you please avoid doing formatting only changes?

File: torch/csrc/jit/passes/shape_analysis.cpp
Comment (apaszke) [inline torch/csrc/jit/passes/shape_analysis.cpp]: Why?

File: torch/csrc/jit/passes/specialize_autogradzero.cpp
Comment (apaszke) [inline torch/csrc/jit/passes/specialize_autogradzero.cpp]: This is unnecessary. We never use `Optional[Tensor]` for those inputs, so it will never get triggered. It might actually be a good idea to get rid of `AutogradZeroTensorType` though (in favor of optionals).

**Review by apaszke — CHANGES_REQUESTED** [review]

File: torch/csrc/jit/fuser/fallback.cpp
Comment (t-vi) [reply torch/csrc/jit/fuser/fallback.cpp]: Oh oh. :+1:

File: torch/csrc/jit/operator.cpp
Comment (t-vi) [reply torch/csrc/jit/operator.cpp]: Well, NoneType isn't in a Subtype relation with OptionalType, so we won't find signatures. Before we had an OptionalType going in.

File: torch/csrc/jit/passes/shape_analysis.cpp
Comment (t-vi) [reply torch/csrc/jit/passes/shape_analysis.cpp]: No it's not. I check for NoneType now to avoid propagating a None outside.
I try to. Here I actually messed up the formatting before and just re-formatting the changed line would have made things worse.

File: torch/csrc/jit/operator.cpp
Comment (apaszke) [reply torch/csrc/jit/operator.cpp]: Hmm ok, I guess we could change it? Those conditionals are worrying for me. I'm not really sure how our type hierarchy looks right now, and if it wouldn't mess up some things. cc @zdevito @eellison @wanchaol

File: torch/csrc/jit/passes/shape_analysis.cpp
Comment (t-vi) [reply torch/csrc/jit/passes/shape_analysis.cpp]: During analysis, we may have None going in when we specialize the graph. This will be in a branch that is later pruned / not executed, but it'll screw the analysis.
I'll add this as a code comment.

File: torch/csrc/jit/passes/shape_analysis.cpp
Comment (t-vi) [reply torch/csrc/jit/passes/shape_analysis.cpp]: Sorry!

File: torch/csrc/jit/passes/specialize_autogradzero.cpp
Comment (t-vi) [reply torch/csrc/jit/passes/specialize_autogradzero.cpp]: Well. I get failures without it. :confused:

Comment (apaszke) [reply torch/csrc/jit/passes/specialize_autogradzero.cpp]: What kinds of failures? I would really like us to understand what's happening instead of applying patches blindly

File: torch/csrc/jit/passes/shape_analysis.cpp
Comment (apaszke) [reply torch/csrc/jit/passes/shape_analysis.cpp]: Why would it screw the analysis? If we know it will deleted later, why does it matter?

File: torch/csrc/jit/operator.cpp
Comment (t-vi) [reply torch/csrc/jit/operator.cpp]: I added it because looked like it's intentional: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/jit_type.h#L221

File: torch/csrc/jit/passes/shape_analysis.cpp
Comment (t-vi) [reply torch/csrc/jit/passes/shape_analysis.cpp]: Well, the code following the unwrap will need a containing type. So we'll hit asserts if we continue with None.

File: torch/csrc/jit/passes/specialize_autogradzero.cpp
Comment (t-vi) [reply torch/csrc/jit/passes/specialize_autogradzero.cpp]: So what happens is for Optional tensor arguments and None inputs, the `ArgumentInfo` (`argument_spec.h`) will have `defined=false`. If now the backward graph takes an `OptionalType` argument, we specialize to `None` (instead of `AutogradZeroTensorType`).
But this then confuses the above if-statement, which differentiates between not-passed-in Tensors and passed-in-Tensors, until now by using `AutogradZeroTensorType`.
In a way, we're involuntarily taking a tiny step towards getting rid of `AutogradZeroTensorType` (which would involve making the replacement in the `ArgumentSpec` unconditional and then one could probably remove the test for `AutogradZeroTensorType` above).

**Comment by t-vi** [comment]
Given the rather fundamental nature of the shortcomings of this approach, I'd have a tendency to drop it.

**Comment by eellison** [comment]
Is this the relevant PR for specializing Optional tensors as graph input?

**Comment by t-vi** [comment]
Yes.

File: test/test_jit.py
Comment (eellison) [inline test/test_jit.py]: fwiw you can write `next` instead of `list(...)[0]`

File: test/test_jit.py
Comment (eellison) [inline test/test_jit.py]: doesn't look like this g is being used

File: torch/csrc/jit/argument_spec.h
Comment (eellison) [inline torch/csrc/jit/argument_spec.h]: Could you add a comment here?

File: torch/csrc/jit/ir.cpp
Comment (eellison) [inline torch/csrc/jit/ir.cpp]: Good fix!

File: torch/csrc/jit/passes/peephole.cpp
Comment (eellison) [inline torch/csrc/jit/passes/peephole.cpp]: This works and we probably don't want to further complicate the PR... I think the more general case is in change constant propagation so that it knows `prim::AutogradZero` is a constant, but i'm not sure how much that would buy us in optimization anyways. you can ignore this comment lol

File: torch/csrc/jit/passes/shape_analysis.cpp
Comment (eellison) [inline torch/csrc/jit/passes/shape_analysis.cpp]: Nit: Could you use the mustBeNone api here instead ?

File: torch/csrc/jit/passes/shape_analysis.cpp
Comment (eellison) [inline torch/csrc/jit/passes/shape_analysis.cpp]: Nit: Maybe rewrite `!node->input()->type()->cast<NoneType>` so that it follows a little more closely with your previous check

File: torch/csrc/jit/operator.cpp
Comment (eellison) [reply torch/csrc/jit/operator.cpp]: I think this is fine as well but maybe @zdevito should weigh in

**Review by eellison — COMMENTED** [review]
Looks good to me, just a few minor nits. I think someone with a better knowledge of the specialize autogradzero pass should review before we land @apaszke @zdevito

File: torch/csrc/jit/ir.cpp
Comment (t-vi) [reply torch/csrc/jit/ir.cpp]: :slightly_smiling_face:

File: torch/csrc/jit/passes/peephole.cpp
Comment (t-vi) [reply torch/csrc/jit/passes/peephole.cpp]: Well, I'm not sure an argument is necessarily a constant (though arguably None is), so constant propagation feels a bit wrong here. But I'm very happy to move it.

File: torch/csrc/jit/passes/shape_analysis.cpp
Comment (t-vi) [reply torch/csrc/jit/passes/shape_analysis.cpp]: Are you certain that that works? I'm happy to do it, but I do rely on your judgement here, that it covers all the cases we need. (And the tests might show.)

File: torch/csrc/jit/operator.cpp
Comment (apaszke) [reply torch/csrc/jit/operator.cpp]: Our `None` story is extremely confusing to be honest. The comment you linked says this as well
> instead NoneType is convertable in schema matching to any Optional[T]

which seems to imply that what you're doing in here should have been already present in the code?

I'd really like to hear what @zdevito thinks about this because I haven't been fully following the `None` changes, but it looks very weird to me.

File: torch/csrc/jit/passes/shape_analysis.cpp
Comment (eellison) [reply torch/csrc/jit/passes/shape_analysis.cpp]: Pretty sure. The only additional case that it is covering is prim::AutoGradZero, in which case we're not gaining anything by continuing with the other two cases right?

File: torch/csrc/jit/passes/peephole.cpp
Comment (eellison) [reply torch/csrc/jit/passes/peephole.cpp]: I think it's good in

File: torch/csrc/jit/passes/shape_analysis.cpp
Comment (apaszke) [inline torch/csrc/jit/passes/shape_analysis.cpp]: When do we emit `unchecked_unwrap_optional`s though? It seems that if they're unchecked then they should never see `NoneType`s as inputs, or they would silently result in UB at runtime? I don't understand what does it mean to "keep the graph consistent". This simply looks wrong

File: torch/csrc/jit/passes/shape_analysis.cpp
Comment (apaszke) [reply torch/csrc/jit/passes/shape_analysis.cpp]: Yeah it looks like a candidate for `mustBeNone()`

File: torch/csrc/jit/passes/shape_analysis.cpp
Comment (t-vi) [reply torch/csrc/jit/passes/shape_analysis.cpp]: I think if we use mustBeNone, we have both covered, so this can be a plain else.

**Comment by eellison** [comment]
I'm not sure that this case will necessarily be hit but just to be safe could you modify the operator of `aten::_unwrap_optionnal` so that it checks for the undefined tensor case ?

File: torch/csrc/jit/argument_spec.h
Comment (apaszke) [reply torch/csrc/jit/argument_spec.h]: Ok, this actually looks wrong to me. Basically when we return an `Optional[Tensor]` from forward, then there are 3 cases to consider for gradients:
1. The result was `None`. It's not a tensor, so it cannot be differentiated.
2. The result was unused/gets no gradient, so it should be treated as a zero tensor.
3. The result was used/gets a gradient, so it is non-zero and we shouldn't optimize it out.

This `if` does not seem to distinguish 1. and 2., which later results in some very awkward code you had to put into AD code, and I would very much like to see those parts gone. A good way to start fixing it is to start differentiating between them in here.

**Review by apaszke — DISMISSED** [review]
I think I finally understand where a lot of confusion is coming from, and why AD has regressed in weird ways. See my comment in `argument_spec.h`.

File: torch/csrc/jit/ir.cpp
Comment (t-vi) [reply torch/csrc/jit/ir.cpp]: Actually, very bad breakage! It throws if we have several outputs. That case needs to be covered in `Value::mustBeNone`.

File: torch/csrc/jit/passes/shape_analysis.cpp
Comment (t-vi) [reply torch/csrc/jit/passes/shape_analysis.cpp]: Well, we know that the if branch containing `unchecked_unwrap_optional` will never be executed. However, the branch is still there during the various JIT passes.

File: torch/csrc/jit/ir.cpp
Comment (eellison) [inline torch/csrc/jit/ir.cpp]: Why was this needed ?

Comment (t-vi) [reply torch/csrc/jit/ir.cpp]: Because of the "bad breakage" above, we can't check for None parameters in `node_->mustBeNone`.
For parameters, you have (in general) more than one output.

**Comment by t-vi** [comment]
So after merging master (including the `ArgumentSpecCreator`), I do distinguish between undefined tensor and None by having an extra flag `is_none_` instead of abusing `defined_`. As the `is_tensor_` is dropped, I added a new `is_nontensor_optional_` (so that zeroing the struct will cover the tensor case).
I actually have the non-tensor case (that used to be part of #18697) here, maybe they're better here.

Note that before this patch, `Optional[Tensor]` hadn't been specialized at all, now we specialize into cases for the various dimensions and `None`. On the other hand `Optional[Tuple[...]]` and `Optional[Class[...]]` only looked at in terms of None / Value, not in terms of their contents (but that wasn't the case before, either, and I don't want to complicate things even more).

File: torch/csrc/jit/passes/specialize_autogradzero.cpp
Comment (t-vi) [reply torch/csrc/jit/passes/specialize_autogradzero.cpp]: This is gone now with better None vs. undefined handling.

File: torch/csrc/jit/argument_spec.h
Comment (t-vi) [reply torch/csrc/jit/argument_spec.h]: This is now fixed to distinguish between all three. :)

File: torch/csrc/jit/passes/shape_analysis.cpp
Comment (t-vi) [reply torch/csrc/jit/passes/shape_analysis.cpp]: Done now with the amended mustBeNone.

**Comment by t-vi** [comment]
So I think we end up with the same thing in the `_jit_pass_complete_shape_analysis`. There was a (pre-existing) bug in the ArgumentSpec test that it compared a `std::move`d  spec to a fresh one.
I could split out the removal of CompleteArgumentSpec if that's preferable. It's not strictly needed, but I felt bad having two differently working ArgumentSpec/CompleteArgumentSpec around when I needed to touch the first.

File: test/cpp/jit/test_argument_spec.h
Comment (zdevito) [inline test/cpp/jit/test_argument_spec.h]: Can you use `jit::compile(str)` to create the graph? We are trying to kill SymbolicVariable.

File: torch/csrc/jit/argument_spec.cpp
Comment (zdevito) [inline torch/csrc/jit/argument_spec.cpp]: comment is copy/paste error.

File: torch/csrc/jit/argument_spec.h
Comment (zdevito) [inline torch/csrc/jit/argument_spec.h]: The way this is encoded is pretty weird. It would be clearer as something like:

// Argspec has:

std::vector<ArgumentInfo> tensor_args; // tensor information recorded
std::vector<bool> optional_presence; // for each optional type true if present, false if not

```
// Related Argspec commands are:
SPECIALIZE_OPTIONAL_TENSOR, // push present/not-present to optional_presense, and if present, add the tensor information to tensor_args
SPECIALIZE_TENSOR, // push tensor information to tensor_args
SPECIALIZE_OPTIONAL, // push present/not-present to optional_presense, 
```

File: torch/csrc/jit/argument_spec.h
Comment (zdevito) [inline torch/csrc/jit/argument_spec.h]: This code needs to be kept. It is being used in the TVM backend.

File: torch/csrc/jit/operator.cpp
Comment (zdevito) [reply torch/csrc/jit/operator.cpp]: This code cannot be modified in this way. NoneType values should not appear in the graph _anywhere_ after the compiler has run. All None's need to be specialized to a particular type otherwise logic related to unwrapping options will no longer work.  Allow None here will cause type variable matching to fail.

File: torch/csrc/jit/operator.cpp
Comment (zdevito) [inline torch/csrc/jit/operator.cpp]: same here

File: torch/csrc/jit/passes/peephole.cpp
Comment (zdevito) [inline torch/csrc/jit/passes/peephole.cpp]: I can't follow this code after a cursory reading. This needs simplification or comments.

File: torch/csrc/jit/passes/shape_analysis.cpp
Comment (zdevito) [inline torch/csrc/jit/passes/shape_analysis.cpp]: Input to unchecked_unwrap_optional should _always_ be an optional type. Things will break if it isn't

File: torch/csrc/jit/passes/shape_analysis.cpp
Comment (zdevito) [inline torch/csrc/jit/passes/shape_analysis.cpp]: same here

File: torch/csrc/jit/argument_spec.cpp
Comment (zdevito) [inline torch/csrc/jit/argument_spec.cpp]: This is the source of the problems with NoneType. Explicit None's are represented as:

```
a : Optional[int] = prim::Constant()
```
in the graph. NoneType is only used as an intermediate in the compiler. Instead of trying to change the input to NoneType. Instead I suggest disconnecting the input that must be none from its uses and replacing it with the appropriate None constant. This will require changing the API here:

ArgumentSpecCreator::specializeTypes(Graph& graph, ...) rather than simply returning new types. This change is fine to make, and will end up helping changes that I may need to do to argspec in the future.

**Review by zdevito — DISMISSED** [review]
This is a really cool PR, but there are a couple of changes that introduce subtle bugs in the type system. See inline comments for how to avoid them.

**Comment by t-vi** [comment]
Thanks @zdevito, it'll make a much shinier patch once I implement your advice.

So one thing I'd need a yes/no on is: Do you want `_jit_pass_complete_shape_analysis` to move to `ArgumentSpecCreator` (the "no" option being to continue to use `CompleteArgumentSpec`)?

File: torch/csrc/jit/argument_spec.h
Comment (t-vi) [reply torch/csrc/jit/argument_spec.h]: So elegant! Thanks!

File: torch/csrc/jit/passes/peephole.cpp
Comment (t-vi) [reply torch/csrc/jit/passes/peephole.cpp]: It's unneeded now.

File: torch/csrc/jit/passes/shape_analysis.cpp
Comment (t-vi) [reply torch/csrc/jit/passes/shape_analysis.cpp]: Well, no. If we specialize optionals, it might be element type instead.

File: torch/csrc/jit/passes/shape_analysis.cpp
Comment (t-vi) [reply torch/csrc/jit/passes/shape_analysis.cpp]: Again, it might be element (sub-) type after specialization.

File: torch/csrc/jit/argument_spec.cpp
Comment (t-vi) [reply torch/csrc/jit/argument_spec.cpp]: This has been the thing I've been missing. Thank you so much!

The remaining not so nice thing here is that I don't know how to specialize Optionals to a Typed-None-Constant inside tuples and objects, because I cannot use `replaceAllUses`.
My tentative conclusion is that I cannot do this currently unless I can add some attribute to the optional type to indicate that it is actually a Typed-None constant.
What could be done is to see which uses are unpacking this, but that's really ugly and probably would remain incomplete.

File: torch/csrc/jit/passes/shape_analysis.cpp
Comment (t-vi) [reply torch/csrc/jit/passes/shape_analysis.cpp]: I changed it to test whether the input is a subtype of the output and if so, take the input type. This allows `Optional[Tensor]`s that have been specialized to `DimensionedTensorType` to have the more precise output type.

File: torch/csrc/jit/passes/shape_analysis.cpp
Comment (t-vi) [reply torch/csrc/jit/passes/shape_analysis.cpp]: So the None part is gone now, and below, Zach has comments on the same place.

File: torch/csrc/jit/ir.cpp
Comment (t-vi) [reply torch/csrc/jit/ir.cpp]: Gone now. :)

File: torch/csrc/jit/passes/shape_analysis.cpp
Comment (t-vi) [reply torch/csrc/jit/passes/shape_analysis.cpp]: They don't see None anymore.

**Comment by t-vi** [comment]
So I think this is much better now. I left open the two things where I think I might not be completely following the advice to the letter (in the unwrapping) or see a limitation.

I left the `_jit_pass_complete_shape_analysis` with using the new mechanism and left `CompleteArgumentSpec` unchanged (except fixing one test).

**Comment by zdevito** [comment]
@t-vi I've been swamped this week, so I haven't gotten to look at this change again, but I just wanted to let you know I haven't forgotten about and will try to get to it as soon as I can.

File: torch/csrc/jit/passes/shape_analysis.cpp
Comment (zdevito) [reply torch/csrc/jit/passes/shape_analysis.cpp]: I think this is ok now.

**Review by zdevito — COMMENTED** [review]
This looks good! Thanks for the patience. I just have one minor question about the autogradzero optional, and one suggestion, otherwise this is ready to go.

File: torch/csrc/jit/argument_spec.cpp
Comment (t-vi) [reply torch/csrc/jit/argument_spec.cpp]: So for an earlier iteration @apaszke had the example of a function returning an `Optional[Tensor]`. On the other hand, if we now specialize `Optional[Tensor]`, we might not need it...

Comment (t-vi) [reply torch/csrc/jit/argument_spec.cpp]: Seems OK without, so I replaced it with an assert.

File: torch/csrc/jit/argument_spec.cpp
Comment (t-vi) [reply torch/csrc/jit/argument_spec.cpp]: I fixed that, thanks!

**Comment by t-vi** [comment]
If there is anything I can do to move this forward, I'd appreciate if you gave me a shout. Thank you!

File: torch/csrc/jit/argument_spec.cpp
Comment (eellison) [inline torch/csrc/jit/argument_spec.cpp]: copy pasta from above

**Review by eellison — APPROVED** [review]
Looks like the comment has been addressed

File: torch/csrc/jit/init.cpp
Comment (wanchaol) [inline torch/csrc/jit/init.cpp]: seems like this switching from completeArgSpec to only propagate dimension tensor information rather than complete tensors. I guess shape analysis works fine in this case, but does this means that we are missing the complete tensor type information? I would imagine if the following passes need the complete information then they could not find it.

Comment (t-vi) [reply torch/csrc/jit/init.cpp]: Wouldn't the comment and code below this get back to CompleteTensorType and propagate those?

**Comment by t-vi** [comment]
So I'm not super-happy with not getting a specialized graph for the failed script:
https://github.com/pytorch/pytorch/blob/dcbfa6759486567ee0e3de76c96c083340fad6d8/test/test_jit.py#L5566-L5571

File: torch/csrc/jit/init.cpp
Comment (wanchaol) [reply torch/csrc/jit/init.cpp]: oops you are right.. I was thinking `incompleteInferTypeFrom` does not get the complete information..

**Review by wanchaol — APPROVED** [review]
lgtm

**Comment by t-vi** [comment]
So now I'm happy with the PyTorch test, but the caffe JIT CI tests are acting up. :(

**Comment by t-vi** [comment]
So finally it's all green again. :)

**Comment by eellison** [comment]
oops didnt see u had imported it @wanchaol

**Pull request closed without merging**
